File: runner/src/coval_bench/providers/stt/deepgram.py
# Copyright 2026 The Coval Benchmarks Authors
# SPDX-License-Identifier: Apache-2.0

"""Deepgram real-time STT provider.

Supports models: default (nova-2), nova-2, nova-3, flux-general-en.
Wire protocol: WebSocket, wss://api.deepgram.com/v1/listen
Auth: Authorization: Token <key>
Close: {"type": "CloseStream"}
"""

from __future__ import annotations

import asyncio
import json
import time  # monotonic clock — wall-clock can step on NTP sync
from typing import Any

import structlog
import websockets.asyncio.client as ws_client
from pydantic import SecretStr

from coval_bench.providers.base import STTProvider, TranscriptionResult

logger = structlog.get_logger(__name__)

_VALID_MODELS = ("default", "nova-2", "nova-3", "flux-general-en")


class DeepgramProvider(STTProvider):
    """Deepgram streaming STT provider."""

    def __init__(self, api_key: SecretStr, model: str = "default") -> None:
        if model not in _VALID_MODELS:
            raise ValueError(f"Invalid Deepgram model {model!r}. Valid: {_VALID_MODELS}")
        self._api_key = api_key
        self._model = model

    @property
    def name(self) -> str:
        if self._model == "default":
            return "deepgram"
        return f"deepgram-{self._model}"

    @property
    def model(self) -> str:
        return self._model

    # ------------------------------------------------------------------
    # URL building
    # ------------------------------------------------------------------

    def _build_websocket_url(self, sample_rate: int, channels: int) -> str:
        if self._model == "flux-general-en":
            # Without interim_results+no_delay the v2/listen preview endpoint
            # buffers transcripts until CloseStream, leaving ttft unmeasurable.
            return (
                "wss://api.preview.deepgram.com/v2/listen"
                "?model=flux-general-en&sample_rate=16000&encoding=linear16"
                "&interim_results=true&no_delay=true"
            )
        url = (
            f"wss://api.deepgram.com/v1/listen"
            f"?sample_rate={sample_rate}"
            f"&encoding=linear16"
            f"&channels={channels}"
            f"&interim_results=true"
            f"&vad_events=true"
            f"&no_delay=true"
        )
        if self._model != "default":
            url += f"&model={self._model}"
        return url

    # ------------------------------------------------------------------
    # Public interface
    # ------------------------------------------------------------------

    async def measure_ttft(
        self,
        audio_data: bytes,
        channels: int,
        sample_width: int,
        sample_rate: int,
        realtime_resolution: float = 0.1,
        audio_duration: float | None = None,
    ) -> TranscriptionResult:
        result = TranscriptionResult(provider=self.name, vad_events_count=0)
        total_start = time.monotonic()

        try:
            url = self._build_websocket_url(sample_rate, channels)
            headers = {"Authorization": f"Token {self._api_key.get_secret_value()}"}

            async with ws_client.connect(url, additional_headers=headers) as ws:
                send_task = asyncio.create_task(
                    self._send_audio(
                        ws,
                        audio_data,
                        channels,
                        sample_width,
                        sample_rate,
                        result,
                        realtime_resolution,
                    )
                )
                recv_task = asyncio.create_task(self._receive(ws, result))
                await asyncio.gather(send_task, recv_task, return_exceptions=True)

        except Exception as exc:
            logger.exception("deepgram measure_ttft failed", error=str(exc))
            result.error = str(exc)

        result.total_time = time.monotonic() - total_start
        return result

    # ------------------------------------------------------------------
    # Send helpers
    # ------------------------------------------------------------------

    async def _send_audio(
        self,
        ws: Any,
        audio_data: bytes,
        channels: int,
        sample_width: int,
        sample_rate: int,
        result: TranscriptionResult,
        realtime_resolution: float,
    ) -> None:
        byte_rate = sample_width * sample_rate * channels
        data = audio_data
        first_chunk = True
        try:
            while data:
                chunk_size = int(byte_rate * realtime_resolution)
                chunk, data = data[:chunk_size], data[chunk_size:]
                if first_chunk:
                    result.audio_start_time = time.monotonic()
                    first_chunk = False
                await ws.send(chunk)
                await asyncio.sleep(realtime_resolution)
            await ws.send(json.dumps({"type": "CloseStream"}))
        except Exception as exc:
            logger.exception("deepgram send error", error=str(exc))
            raise

    # ------------------------------------------------------------------
    # Receive helpers
    # ------------------------------------------------------------------

    async def _receive(self, ws: Any, result: TranscriptionResult) -> None:
        final_segments: list[str] = []
        flux_latest: str = ""
        last_final_time: float | None = None

        try:
            async for raw in ws:
                if isinstance(raw, bytes):
                    continue

                msg: dict[str, Any] = json.loads(raw)
                now = time.monotonic()
                msg_type: str = msg.get("type", "")

                # VAD events
                if msg_type == "SpeechStarted":
                    if result.audio_start_time is not None:
                        elapsed = now - result.audio_start_time
                        if result.vad_first_detected is None:
                            result.vad_first_detected = elapsed
                            result.vad_first_event_content = str(msg)
                    result.vad_events_count = (result.vad_events_count or 0) + 1
                    continue

                if msg_type in ("SpeechEnded", "Metadata", "Connected", "TurnInfo"):
                    continue

                transcript = self._extract_transcript(msg)
                if not transcript:
                    continue

                # TTFT
                if result.ttft_seconds is None and result.audio_start_time is not None:
                    result.ttft_seconds = now - result.audio_start_time
                    result.first_token_content = (
                        transcript[:30] + "..." if len(transcript) > 30 else transcript
                    )

                result.partial_transcripts.append(transcript)

                if self._model == "flux-general-en":
                    flux_latest = transcript
                    # flux emits rolling updates; every transcript update is the
                    # latest "final" view of what was said — track the last one.
                    last_final_time = now
                elif self._model in ("nova-2", "nova-3"):
                    if msg.get("speech_final"):
                        final_segments.append(transcript)
                        last_final_time = now
                else:
                    # default model — treat is_final as final
                    if msg.get("is_final"):
                        final_segments.append(transcript)
                        last_final_time = now

        except Exception as exc:
            logger.exception("deepgram receive error", error=str(exc))

        if last_final_time is not None and result.audio_start_time is not None:
            result.audio_to_final_seconds = last_final_time - result.audio_start_time

        # Build complete transcript
        if self._model == "flux-general-en":
            result.complete_transcript = flux_latest.strip() or None
        elif final_segments:
            result.complete_transcript = " ".join(final_segments).strip()
        elif result.partial_transcripts:
            result.complete_transcript = max(result.partial_transcripts, key=len).strip() or None

        if result.complete_transcript:
            result.transcript_length = len(result.complete_transcript)
            result.word_count = len(result.complete_transcript.split())

    # ------------------------------------------------------------------
    # Transcript extraction
    # ------------------------------------------------------------------

    def _extract_transcript(self, msg: dict[str, Any]) -> str:
        # Both the standard endpoint (v1/listen) and the flux preview endpoint
        # (v2/listen) return Deepgram's standard Results message shape:
        #   {"type": "Results", "channel": {"alternatives": [{"transcript": "..."}]}}
        # An earlier implementation read a top-level "transcript" key for flux,
        # but that key does not exist in a Results message — causing silent NULL
        # TTFT for flux-general-en on every run.
        try:
            channel: dict[str, Any] = msg.get("channel", {})
            alternatives: list[dict[str, Any]] = channel.get("alternatives", [])
            if not alternatives:
                return ""
            alt = alternatives[0]
            words: list[dict[str, Any]] = alt.get("words", [])
            if words:
                parts = []
                for w in words:
                    text = (
                        str(w.get("punctuated_word", "")).strip() or str(w.get("word", "")).strip()
                    )
                    if text:
                        parts.append(text)
                if parts:
                    return " ".join(parts)
            return str(alt.get("transcript", "")).strip()
        except (KeyError, IndexError, TypeError):
            return ""

File: web/lib/stt/deepgram.ts
import WebSocket from "@/lib/stt/ws";

// 16 kHz × 16-bit × 1 channel × 0.1s = 3200 bytes. Same for nova and flux —
// both endpoints accept linear16/16k/mono streamed in 100ms PCM frames.
const DEEPGRAM_PCM_CHUNK_BYTES = 3_200;

export type DeepgramResult = {
  transcript: string;
  ttftMs: number | null;
  audioToFinalMs: number;
};

export async function callDeepgram(
  pcm: ArrayBuffer,
  model: string,
  apiKey: string,
): Promise<DeepgramResult> {
  if (model === "flux-general-en") {
    return callDeepgramFlux(pcm, apiKey);
  }
  // nova-2 / nova-3 — v1 WebSocket protocol, mirrors runner's
  // DeepgramProvider._build_websocket_url so playground TTFT/audio-to-final
  // numbers are comparable with the benchmark dashboard.
  return callDeepgramNova(pcm, model, apiKey);
}

function callDeepgramNova(
  pcm: ArrayBuffer,
  model: string,
  apiKey: string,
): Promise<DeepgramResult> {
  const params = new URLSearchParams({
    sample_rate: "16000",
    encoding: "linear16",
    channels: "1",
    interim_results: "true",
    vad_events: "true",
    no_delay: "true",
    model,
  });
  const url = `wss://api.deepgram.com/v1/listen?${params.toString()}`;
  const buf = Buffer.from(pcm);

  return new Promise((resolve, reject) => {
    const ws = new WebSocket(url, {
      headers: { Authorization: `Token ${apiKey}` },
    });

    let settled = false;
    let sentAudio = false;
    let t0 = 0;
    let ttftMs: number | null = null;
    const finalSegments: string[] = [];
    let lastInterim = "";
    let lastFinalTime = 0;

    const settle = (fn: () => void) => {
      if (settled) return;
      settled = true;
      fn();
    };

    const timer = setTimeout(() => {
      ws.close();
      settle(() => reject(new Error("Deepgram Nova: timeout")));
    }, 55_000);
    const clearTimer = () => clearTimeout(timer);

    ws.on("open", () => {
      t0 = performance.now();
      sentAudio = true;
      for (let i = 0; i < buf.length; i += DEEPGRAM_PCM_CHUNK_BYTES) {
        const slice = buf.subarray(i, Math.min(i + DEEPGRAM_PCM_CHUNK_BYTES, buf.length));
        if (slice.length > 0) ws.send(slice);
      }
      ws.send(JSON.stringify({ type: "CloseStream" }));
    });

    ws.on("message", (raw) => {
      const text = Buffer.isBuffer(raw) ? raw.toString("utf8") : String(raw);
      if (!text.trimStart().startsWith("{")) return;

      let msg: {
        type?: string;
        description?: string;
        speech_final?: boolean;
        is_final?: boolean;
        transcript?: string;
        channel?: {
          alternatives?: Array<{
            transcript?: string;
            words?: Array<{ word?: string; punctuated_word?: string }>;
          }>;
        };
      };
      try {
        msg = JSON.parse(text) as typeof msg;
      } catch {
        return;
      }

      if (msg.type === "Error") {
        settle(() => {
          clearTimer();
          ws.close();
          reject(new Error(`Deepgram Nova: ${msg.description ?? "error"}`));
        });
        return;
      }

      if (msg.type !== "Results") return;

      const transcript = extractDeepgramTranscript(msg);
      if (!transcript) return;

      if (ttftMs === null && t0 > 0) ttftMs = Math.round(performance.now() - t0);

      if (msg.speech_final) {
        finalSegments.push(transcript);
        lastFinalTime = performance.now();
      } else {
        lastInterim = transcript;
      }
    });

    ws.on("close", () => {
      clearTimer();
      settle(() => {
        if (!sentAudio) {
          reject(new Error("Deepgram Nova: connection closed before audio send"));
          return;
        }
        const transcript =
          finalSegments.length > 0 ? finalSegments.join(" ").trim() : lastInterim.trim();
        const audioToFinalMs = lastFinalTime > 0
          ? Math.round(lastFinalTime - t0)
          : (t0 > 0 ? Math.round(performance.now() - t0) : 0);
        resolve({ transcript, ttftMs, audioToFinalMs });
      });
    });

    ws.on("error", (err) => {
      clearTimer();
      settle(() => reject(err));
    });
  });
}

function callDeepgramFlux(pcm: ArrayBuffer, apiKey: string): Promise<DeepgramResult> {
  const params = new URLSearchParams({
    model: "flux-general-en",
    encoding: "linear16",
    sample_rate: "16000",
  });
  const url = `wss://api.deepgram.com/v2/listen?${params.toString()}`;
  const buf = Buffer.from(pcm);

  return new Promise((resolve, reject) => {
    const ws = new WebSocket(url, {
      headers: { Authorization: `Token ${apiKey}` },
    });

    let settled = false;
    let connectedSeen = false;
    let sentAudio = false;
    let t0 = 0;
    let ttftMs: number | null = null;
    const endTurnTexts: string[] = [];
    let lastTranscript = "";

    const settle = (fn: () => void) => {
      if (settled) return;
      settled = true;
      fn();
    };

    const timer = setTimeout(() => {
      ws.close();
      settle(() => reject(new Error("Deepgram Flux: timeout")));
    }, 55_000);

    const clearTimer = () => clearTimeout(timer);

    ws.on("message", (raw) => {
      const text = Buffer.isBuffer(raw) ? raw.toString("utf8") : String(raw);
      if (!text.trimStart().startsWith("{")) return;

      let msg: {
        type?: string;
        event?: string;
        transcript?: string;
        description?: string;
        code?: string;
        channel?: {
          alternatives?: Array<{
            transcript?: string;
            words?: Array<{ word?: string; punctuated_word?: string }>;
          }>;
        };
      };
      try {
        msg = JSON.parse(text) as typeof msg;
      } catch {
        return;
      }

      switch (msg.type) {
        case "Connected": {
          if (sentAudio) return;
          sentAudio = true;
          connectedSeen = true;
          t0 = performance.now();
          for (let i = 0; i < buf.length; i += DEEPGRAM_PCM_CHUNK_BYTES) {
            const slice = buf.subarray(i, Math.min(i + DEEPGRAM_PCM_CHUNK_BYTES, buf.length));
            if (slice.length > 0) ws.send(slice);
          }
          ws.send(JSON.stringify({ type: "CloseStream" }));
          return;
        }
        case "TurnInfo": {
          const transcript = extractDeepgramTranscript(msg);
          if (transcript) {
            if (ttftMs === null && t0 > 0) ttftMs = Math.round(performance.now() - t0);
            lastTranscript = transcript;
          }
          if (msg.event === "EndOfTurn" && transcript) {
            endTurnTexts.push(transcript);
          }
          return;
        }
        case "Results": {
          const transcript = extractDeepgramTranscript(msg);
          if (transcript) {
            if (ttftMs === null && t0 > 0) ttftMs = Math.round(performance.now() - t0);
            lastTranscript = transcript;
          }
          return;
        }
        case "Error": {
          settle(() => {
            clearTimer();
            ws.close();
            reject(new Error(`Deepgram Flux: ${msg.description ?? msg.code ?? "error"}`));
          });
          return;
        }
        default:
          return;
      }
    });

    ws.on("close", () => {
      clearTimer();
      settle(() => {
        if (!connectedSeen) {
          reject(new Error("Deepgram Flux: connection closed before Connected (check API key)"));
          return;
        }
        const transcript =
          endTurnTexts.length > 0 ? endTurnTexts.join(" ").trim() : lastTranscript.trim();
        const audioToFinalMs = t0 > 0 ? Math.round(performance.now() - t0) : 0;
        resolve({ transcript, ttftMs, audioToFinalMs });
      });
    });

    ws.on("error", (err) => {
      clearTimer();
      settle(() => reject(err));
    });
  });
}

function extractDeepgramTranscript(msg: {
  transcript?: string;
  channel?: {
    alternatives?: Array<{
      transcript?: string;
      words?: Array<{ word?: string; punctuated_word?: string }>;
    }>;
  };
}): string {
  if (typeof msg.transcript === "string" && msg.transcript.trim()) {
    return msg.transcript.trim();
  }
  const alt = msg.channel?.alternatives?.[0];
  if (!alt) return "";
  const words = alt.words ?? [];
  const wordText = words
    .map((w) => (w.punctuated_word ?? w.word ?? "").trim())
    .filter(Boolean)
    .join(" ");
  return wordText || (alt.transcript ?? "").trim();
}
